backend/controllers/cellMove/cellMoveHistory.js

```diff
@@ -0,0 +1,135 @@
+
+const matchesLocationPrefix = (description, locationPrefix) => {
+  if (locationPrefix.length > description.length) return false
+
+  const data = description.substr(0, locationPrefix.length)
+  return data === locationPrefix
+}
+
+const dateTimeFormat = 'YYYY-MM-DDTHH:mm'
+
+const sortByMostEarliestFirst = (left, right) => {
+  const leftDate = moment(left.assignmentDateTime, dateTimeFormat)
+  const rightDate = moment(right.assignmentDateTime, dateTimeFormat)
+
+  if (leftDate.isAfter(rightDate, 'minute')) return 1
+  if (leftDate.isBefore(rightDate, 'minute')) return -1
+
+  return 0
+}
+
+const prisonApiLocationDescription = async (res, whereaboutsApi, locationKey, userCaseLoad) => {
+  const fullLocationPrefix = await whereaboutsApi.getAgencyGroupLocationPrefix(res.locals, userCaseLoad, locationKey)
+  if (fullLocationPrefix) {
+    const locationIdWithSuffix = fullLocationPrefix.locationPrefix
+    return locationIdWithSuffix?.length < 1 ? '' : locationIdWithSuffix.slice(0, -1)
+  }
+  return `${userCaseLoad}-${locationKey}`
+}
+
+module.exports = ({ prisonApi, whereaboutsApi }) => async (req, res) => {
+  if (!req?.query?.date) return res.redirect('/change-someones-cell/recent-cell-moves')
+
+  const { date, locationId, reason } = req.query
+
+  const {
+    user: { activeCaseLoad },
+  } = res.locals
+
+  const filterByLocationPrefix =
+    locationId &&
+    (await prisonApiLocationDescription(res, whereaboutsApi, req.query.locationId, activeCaseLoad.caseLoadId))
+
+  const currentUserLocations = await whereaboutsApi.searchGroups(res.locals, activeCaseLoad.caseLoadId)
+
+  const locations = currentUserLocations.map(locationData => ({ text: locationData.name, value: locationData.key }))
+
+  const cellMoveHistory = (await prisonApi.getHistoryByDate(res.locals, {
+    assignmentDate: date,
+  })).filter(
+    item =>
+      item.agencyId === activeCaseLoad.caseLoadId &&
+      (!filterByLocationPrefix || matchesLocationPrefix(item.description, filterByLocationPrefix)) &&
+      (!reason || item.assignmentReason === reason)
+  )
+  const usernames = [...new Set(cellMoveHistory.map(cellMove => cellMove.movementMadeBy))]
+  const offenderNos = [...new Set(cellMoveHistory.map(cellMove => cellMove.offenderNo))]
+  const cellMoveTypes = await prisonApi.getCellMoveReasonTypes(res.locals)
+
+  const staffMembers = await Promise.all(usernames.map(staffId => prisonApi.getStaffDetails(res.locals, staffId)))
+
+  const cellMoveReasons = cellMoveTypes.map(subType => ({
+    value: subType.code,
+    text: subType.description,
+  }))
+
+  const offenders = await prisonApi.getPrisoners(
+    { ...res.locals, requestHeaders: { 'page-offset': 0, 'page-limit': offenderNos.length } },
+    { offenderNos }
+  )
+
+  const bookingIds = (offenders && [...new Set(offenders.map(o => o.latestBookingId))]) || []
+
+  const cellHistoryByOffenderNo = await Promise.all(
+    bookingIds.map(bookingId =>
+      prisonApi
```

Comment:
We really ought to do something in `prison-api` to get this information, because this is really going to have an impact on performance. Not now, but if we add a tech ticket we can sort it out - I assume a similar thing is done elsewhere.